lock-api-objects/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.logsafe.Preconditions;
+
+public final class ClientLockWatchEventLogImpl implements ClientLockWatchEventLog {
+    private final ProcessingVisitor processingVisitor = new ProcessingVisitor();
+    private final NewLeaderVisitor newLeaderVisitor = new NewLeaderVisitor();
+    private final ConcurrentSkipListMap<Long, LockWatchEvent> eventLog;
+    private IdentifiedVersion identifiedVersion;
+    private LockWatchStateUpdate.Snapshot seed = failedSnapshot(UUID.randomUUID());
+
+    private ClientLockWatchEventLogImpl() {
+        identifiedVersion = IdentifiedVersion.of(UUID.randomUUID(), Optional.empty());
+        eventLog = new ConcurrentSkipListMap<>();
+    }
+
+    @Override
+    public IdentifiedVersion getLatestKnownVersion() {
+        return identifiedVersion;
+    }
+
+    @Override
+    public IdentifiedVersion processUpdate(LockWatchStateUpdate update) {
+        if (update.logId().equals(identifiedVersion.id())) {
+            update.accept(processingVisitor);
+            return identifiedVersion;
+        } else {
+            update.accept(newLeaderVisitor);
+            return identifiedVersion;
+        }
+    }
+
+    @Override
+    public TransactionsLockWatchEvents getEventsForTransactions(
+            Map<Long, Long> timestampToVersion,
+            IdentifiedVersion version) {
+        if (!version.id().equals(identifiedVersion.id())) {
+            return TransactionsLockWatchEvents.failure(seed);
+        }
+
+        Long latestVersion = Collections.max(timestampToVersion.values());
+        ConcurrentSkipListMap<Long, LockWatchEvent> logSnapshot = eventLog.clone();
+
+        if (logSnapshot.isEmpty()) {
+            return TransactionsLockWatchEvents.success(ImmutableList.of(), timestampToVersion);
+        }
+
+        long oldestVersion = version.version().orElseGet(logSnapshot::firstKey);
+        Preconditions.checkArgument(oldestVersion <= latestVersion, "startVersion should be before endVersion");
+
+        long startKey = logSnapshot.ceilingKey(oldestVersion);
+        long endKey = logSnapshot.floorKey(latestVersion);
+
+        return TransactionsLockWatchEvents.success(
+                new ArrayList<>(logSnapshot.subMap(startKey, endKey).values()),
+                timestampToVersion);
+    }
+
+    private synchronized void processSuccess(LockWatchStateUpdate.Success success) {
```

Comment:
determining whether synchronized is necessary is one of the remaining pieces of work here.